http-middleware/security.go

```diff
@@ -0,0 +1,224 @@
+package httpmiddleware
+
+import (
+	"context"
+	"net/http"
+
+	http_model "github.com/SKF/go-utility/http-model"
+	"github.com/SKF/go-utility/jwk"
+	"github.com/SKF/go-utility/jwt"
+	"github.com/SKF/go-utility/log"
+	"github.com/SKF/proto/common"
+
+	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
+	"go.opencensus.io/trace"
+)
+
+type claimsContextKey struct{}
+
+const (
+	HeaderAuthorization = "Authorization"
+)
+
+// AuthenticateMiddleware retrieves the security configuration for the matched route
+// and handles Access Token validation and stores the token claims in the request context.
+func AuthenticateMiddleware(keySetURL string) mux.MiddlewareFunc {
+	jwk.KeySetURL = keySetURL
+	if err := jwk.RefreshKeySets(); err != nil {
+		log.
+			WithError(err).
+			Error("Couldn't refresh JWKeySets")
+	}
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			ctx, span := trace.StartSpan(req.Context(), "Authenticator")
+			defer span.End()
+			*req = *req.WithContext(ctx)
+
+			logFields := log.
+				WithField("method", req.Method).
+				WithField("url", req.URL.String())
+
+			secConfig := lookupSecurityConfig(req)
+			if secConfig.accessTokenHeader != "" {
+				if err := handleAccessToken(req, secConfig.accessTokenHeader); err != nil {
+					logFields.WithError(err).Warn("User is not authorized")
+					http_model.WriteJSONResponse(w, http.StatusUnauthorized, http_model.ErrResponseUnauthorized)
+					return
+				}
+			}
+
+			span.End()
```

Comment:
Is this span needed when you have the defer above?

Reply:
yep, so the reason for it to be there is that otherwise we will include `next.ServeHTTP(w, req)` in the same span as the authorizer call